securedrop/tests/test_source.py

```diff
@@ -318,3 +318,24 @@ def test_login_with_overly_long_codename(self, mock_hash_codename):
             self.assertFalse(mock_hash_codename.called,
                              "Called hash_codename for codename w/ invalid "
                              "length")
+
+    @patch('source.app.logger.warning')
+    def test_generate_unique_codename(self, mock_warning):
+        codename = source.generate_unique_codename()
+        self.assertGreater(len(codename), 0)
+        mock_warning.assert_not_called()
+
+    @patch('source.crypto_util.genrandomid')
+    @patch('source.app.logger.warning')
+    def test_generate_unique_codename_too_long(self, mock_warning, mock_genrandomid):
+        self.too_long = True
+        def genrandomid(num_words):
+            if self.too_long:
+                self.too_long = False
+                return 'O' * Source.MAX_CODENAME_LEN * 2
+            else:
+                return 'not so random'
+        mock_genrandomid.side_effect = genrandomid
```

Comment:
You could simplify the mock by using an [iterator](https://docs.python.org/dev/library/unittest.mock.html#unittest.mock.Mock.side_effect) instead of a function for `side_effect`, e.g.

```py
    @patch('source.crypto_util.genrandomid')
    @patch('source.app.logger.warning')
    def test_generate_unique_codename_too_long(self, mock_warning, mock_genrandomid):
        mock_genrandomid.side_effect = [
            'O' * Source.MAX_CODENAME_LEN * 2,
            'not too long'
        ]
        codename = source.generate_unique_codename()
        mock_warning.assert_called()
        self.assertEqual(codename, 'not too long')
```